### PR TITLE
EG-1856: Allow users to specify Dockerform ports

### PR DESCRIPTION
**Pull request by palade**
Dockerform allocates predefined port numbers instead of using the ones specified in the `node.conf`. Users should be allowed to specify those ports

### CODE REVIEW

File: cordformation/src/main/kotlin/net/corda/plugins/RpcSettings.kt [hunk 1]
```diff
@@ -34,6 +39,9 @@ class RpcSettings {
      * RPC admin address for the node (necessary if [useSsl] is false or unset).
      */
     fun adminAddress(value: String) {
+        try {
+            adminPort = URI(null, value, null, null, null).port
+        } catch (ex: URISyntaxException) { }
```
Comment (chrisr3) [inline cordformation/src/main/kotlin/net/corda/plugins/RpcSettings.kt:44]: Same as above. And maybe make a function that both properties can use for this?

File: cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt [hunk 2]
```diff
@@ -99,7 +99,7 @@ open class Dockerform @Inject constructor(objects: ObjectFactory) : Baseform(obj
                             "$nodeBuildDir/additional-node-infos:/opt/corda/additional-node-infos",
                             "$nodeBuildDir/drivers:/opt/corda/drivers"
                     ),
-                    "ports" to listOf(it.rpcPort, it.config.getInt("sshd.port")),
+                    "ports" to listOf(it.rpcPort(), it.config.getInt("sshd.port")),
```
[the remark this answers is not in the document]
Comment (palade) [reply cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt:102]: > Does `sshd.port` need to be different for each node too?

Not necessarily, but a user may want to see the port number that he set in the `build.conf`

**Comment by palade** [comment]
> Looks OK. Is there any way that your new test can also check whether the RPC ports have been set correctly? After all, that _is_ the point of this PR.

Added a unit test to check for correct parsing of the RPC  port. Moved the parsing function in separate file and added several unit tests for that function as well

File: cordformation/src/test/kotlin/net/corda/plugins/DockerformTest.kt [hunk 3]
```diff
@@ -111,5 +114,23 @@ class DockerformTest : BaseformTest() {
...
         assertThat(getNodeCordappJar(bankOfCordaNodeName, cordaFinanceContractsJarName)).isRegularFile()
         assertThat(getNetworkParameterOverrides(bankOfCordaNodeName)).isRegularFile()
+
+        val notaryConfigPath = getNodeConfig(notaryNodeName)
+        assertThat(notaryConfigPath).isRegularFile()
+
+        val notaryConfig = ConfigFactory.parseFile(notaryConfigPath.toFile())
+        assertTrue(notaryConfig.hasPath("rpcSettings.address"))
+        assertTrue(notaryConfig.hasPath("rpcSettings.adminAddress"))
+        assertEquals(ConfigurationUtils.parsePort(notaryConfig.getString("rpcSettings.address")), 10003)
```
Comment (chrisr3) [inline cordformation/src/test/kotlin/net/corda/plugins/DockerformTest.kt:124]: Funny thing - this `assertEquals()` function is really:
```kotlin
assertEquals(expectedValue, actualValue)
```
i.e. the assertion should be written as:
```kotlin
assertEquals(10003, ConfigurationUtils.parsePort(notaryConfig.getString("rpcSettings.address")))
```
(FYI different frameworks put these parameters in different orders. I don't know why....)

File: cordformation/src/test/kotlin/net/corda/plugins/ConfigurationUtilsTest.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+import org.junit.jupiter.api.Test
+import kotlin.test.assertEquals
+
+class ConfigurationUtilsTest {
+
+    @Test
+    fun `check correct port value parsing`() {
+        assertEquals(ConfigurationUtils.parsePort("localhost:10000"), 10000)
```
Comment (chrisr3) [inline cordformation/src/test/kotlin/net/corda/plugins/ConfigurationUtilsTest.kt:11]: Oops, `assertEquals(expectedValue, actualValue)`...

File: cordformation/src/main/kotlin/net/corda/plugins/ConfigurationUtils.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+class ConfigurationUtils {
+
+    companion object {
+
+        fun parsePort(address: String): Int {
+            return try {
+                URI(null, address, null, null, null).port
+            } catch (ex: URISyntaxException) {
+                throw InvalidUserDataException("Invalid host and port syntax for RPC address, expected host:port. Using default value")
```
Comment (chrisr3) [inline cordformation/src/main/kotlin/net/corda/plugins/ConfigurationUtils.kt:15]: This exception will kill Gradle, which is fine. It's just that the message shouldn't therefore say that the default value will be used.

File: cordformation/src/test/kotlin/net/corda/plugins/DockerformTest.kt (hunk 3 above)
Comment (palade) [reply cordformation/src/test/kotlin/net/corda/plugins/DockerformTest.kt:124]: Fixed now :-)

**Pull request merged**